rust/eos-rs/src/api/dungeon_mode/monster.rs

```diff
@@ -1,334 +1,104 @@
-
-    /// Checks if a monster is currently immune to Ground-type moves for reasons other than typing and ability.
-    ///
-    /// This includes checks for Gravity and Magnet Rise.
-    fn has_conditional_ground_immunity(&self) -> bool;
-
-    /// Gets the move target-and-range field when used by a given entity.
-    ///
-    /// The fourth field in the returned tuple seems unused.
-    /// The values in the returned tuple are None, if they are invalid (or we don't know them yet).
-    ///
-    /// See [`Move::get_target_and_range`] for more information.
-    fn get_move_target_and_range(&self, the_move: &Move, is_ai: bool) -> MoveTargetAndRange;
-
-    /// Get the weather, as experienced by the monster.
-    fn get_apparent_weather(&self) -> Option<Weather>;
-
-    /// Checks if the monster has a certain held item.
-    fn has_held_item(&self, item_id: item_catalog::Type) -> bool;
-
-    /// Gets the power of a move, factoring in Ginseng/Space Globe boosts.
-    fn get_move_power(&self, the_move: &Move) -> i32;
-
-    /// Seems to calculate the duration of a volatile status on a monster.
-    ///
-    /// Returns the number of turns for the status condition.
-    ///
-    /// # Arguments
-    /// * `turn_range` - lower & higher ends of the turn range
-    /// * `effects` - flag for whether or not to factor in the Self Curer IQ skill and the
-    ///               Natural Cure ability
-    fn calc_status_duration(&self, turn_range: &[u16; 2], effects: bool) -> i32;
-
-    /// Returns the number of attacks that a monster can do in one turn (1 or 2).
-    ///
-    /// Checks for the abilities Swift Swim, Chlorophyll, Unburden, and for exclusive items.
-    fn get_number_of_attacks(&self) -> i32;
-
-    /// Checks if a monster is levitating (has the effect of Levitate and Gravity is not active).
-    fn is_levitating(&self) -> bool;
-
-    /// Checks if the monster is under the effect of Conversion 2 (its type was changed). Returns
-    /// `None` if the value is invalid.
-    fn is_conversion2_active(&self) -> Option<Conversion2Status>;
-
-    /// Check the type of a move when used by a certain monster. Accounts for special cases
-    /// such as Hidden Power, Weather Ball, the regular attack...
-    fn get_move_type_if_used_by_self(&self, the_move: &Move) -> type_catalog::Type;
+/// Mutable reference type for the info struct for [`DungeonEntity`] objects that are monsters.
+pub struct DungeonMonsterMut<'a>(pub &'a mut ffi::monster, pub &'a mut DungeonEntity);
 
-    /// Returns the animation id to be applied to a monster that has the sleep, napping,
-    /// nightmare or bide status.
-    fn get_sleep_animation_id(&self) -> u8;
-
-    /// Returns true if the monster has the blinded status, or if it is not the leader and is
-    /// holding Y-Ray Specs.
-    fn is_blinded(&self, check_held_item: bool) -> bool;
+/// Essentially [`core::convert::AsRef`].
+/// (Sadly we can't use the `AsRef` trait, because it doesn't allow explicit lifetimes.)
```

Comment:
I'm not convinced that this implementation pattern with `DungeonMonsterRef`/`DungeonMonsterMut` is the best way to implement things.

I think the core issue here is that you want `DungeonMonsterMut` to have access to all the same methods as `DungeonMonsterRef`, but not the other way around. This solution handles this by mocking an `AsRef` implementation. On top of the fact that this is a bit unnatural for struct conversion (as evidenced by the fact that you can't _actually_ implement the real `AsRef`), I think it also impact usability since you're going to need to do `my_DungeonMonsterMut.as_ref().some_DungeonMonsterRef_method()` rather than `my_DungeonMonsterMut.some_DungeonMonsterRef_method()`, right?

What if we used blanket implementations instead? I haven't actually checked if this compiles, but maybe something like:

```rust
pub trait MonsterEntity {
    fn monster(&self) -> &DungeonMonster;
    fn entity(&self) -> &DungeonEntity;

    fn is_special_story_ally(&self) -> bool {
        // use self.monster() instead of self.0
    }
    // etc.
}

pub trait MonsterEntityMut {
    fn monster_mut(&self) -> &mut DungeonMonster;
    fn entity_mut(&self) -> &mut DungeonEntity;

    fn update_move_pp(&mut self, should_consume_pp: bool) {
        // use self.entity_mut() instead of self.1
    }
    // etc.
}

impl<'a> AsRef<DungeonEntity> for DungeonMonsterRef<'a> {...}
impl<'a> AsRef<DungeonMonster> for DungeonMonsterRef<'a> {...}

impl<'a> AsRef<DungeonEntity> for DungeonMonsterMut<'a> {...}
impl<'a> AsRef<DungeonMonster> for DungeonMonsterMut<'a> {...}
impl<'a> AsMut<DungeonEntity> for DungeonMonsterMut<'a> {...}
impl<'a> AsMut<DungeonMonster> for DungeonMonsterMut<'a> {...}

impl<T> MonsterEntity for T where T: AsRef<DungeonEntity> + AsRef<DungeonMonster> {...}
impl<T> MonsterEntityMut for T where T: AsMut<DungeonEntity> + AsMut<DungeonMonster> {...}
```

I guess the generics might cause code bloat though, which maybe is undesirable 🤷.

Reply:
I mean if you look at the diff, that is exactly what I did before. I guess I could revert this to work with traits again.
I'll actually check how RefMut does it, it must face a similiar issue, let's see how they solved it there.

Reply:
I think I'll leave it like this for now, but might revisit this later. It's just the traits were getting very big and the blanket implementations also felt kinda hacky. I guess I could make the `DungeonMonsterMut` additionally Deref into `DungeonMonsterRef`, then it would behave like Rust's own implicit &mut T -> &T conversion, but I'm not sure if that wouldn't be confusing, what do you think?

Reply:
Oh right that doesn't work because Deref can only return real references. I think I'll revert back to the trait solution then.

Reply:
The new implementation LGTM.